File: backend/src/main/java/com/interviewer/service/ProctoringService.java
package com.interviewer.service;

import com.interviewer.dto.ProctorEventRequest;
import com.interviewer.model.InterviewSession;
import com.interviewer.model.ProctorEvent;
import com.interviewer.repository.ProctorEventRepository;
import com.interviewer.repository.SessionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class ProctoringService {

    @Autowired
    private ProctorEventRepository proctorEventRepository;

    @Autowired
    private SessionRepository sessionRepository;

    // Deduction values for each event type
    private static final Map<String, Integer> DEDUCTIONS = Map.of(
            "TAB_SWITCH", 10,
            "NO_FACE", 5,
            "MULTIPLE_FACES", 15,
            "LOOKING_AWAY", 3,
            "LOOKING_DOWN", 2,
            "BACKGROUND_NOISE", 2
    );

    /**
     * Log a proctoring event and update integrity score
     */
    public Map<String, Object> logEvent(ProctorEventRequest request) {
        // Save the event
        ProctorEvent event = new ProctorEvent();
        event.setSessionId(request.getSessionId());
        event.setEventType(request.getEventType());
        event.setSeverity(request.getSeverity());
        event.setDescription(request.getDescription() != null ?
                request.getDescription() :
                getDefaultDescription(request.getEventType()));

        proctorEventRepository.save(event);

        // Update session integrity score
        InterviewSession session = sessionRepository.findById(request.getSessionId())
                .orElseThrow(() -> new RuntimeException("Session not found"));

        int deduction = DEDUCTIONS.getOrDefault(request.getEventType(), 2);
        int newScore = Math.max(0, session.getIntegrityScore() - deduction);
        session.setIntegrityScore(newScore);
        session.setWarningCount(session.getWarningCount() + 1);
        sessionRepository.save(session);

        // Build response
        Map<String, Object> result = new HashMap<>();
        result.put("integrityScore", newScore);
        result.put("warningCount", session.getWarningCount());
        result.put("warning", getWarningMessage(request.getEventType()));
        result.put("deduction", deduction);
        return result;
    }

    /**
     * Get integrity score for a session
     */
    public Map<String, Object> getIntegrityScore(String sessionId) {
        InterviewSession session = sessionRepository.findById(sessionId)
                .orElseThrow(() -> new RuntimeException("Session not found"));

        Map<String, Object> result = new HashMap<>();
        result.put("integrityScore", session.getIntegrityScore());
        result.put("warningCount", session.getWarningCount());
        return result;
    }

    /**
     * Calculate integrity score from events (used for reports)
     */
    public int calculateIntegrity(int tabSwitches, int noFace, int multipleFaces, int lookingAway) {
        int deduction = (tabSwitches * 10) + (noFace * 5) + (multipleFaces * 15) + (lookingAway * 3);
        return Math.max(0, 100 - deduction);
    }

    /**
     * Get proctoring summary for report
     */
    public Map<String, Integer> getProctoringSummary(String sessionId) {
        Map<String, Integer> summary = new HashMap<>();
        summary.put("tabSwitches", (int) proctorEventRepository.countBySessionIdAndEventType(sessionId, "TAB_SWITCH"));
        summary.put("noFaceCount", (int) proctorEventRepository.countBySessionIdAndEventType(sessionId, "NO_FACE"));
        summary.put("multipleFaces", (int) proctorEventRepository.countBySessionIdAndEventType(sessionId, "MULTIPLE_FACES"));
        summary.put("lookingAway", (int) proctorEventRepository.countBySessionIdAndEventType(sessionId, "LOOKING_AWAY"));
        summary.put("lookingDown", (int) proctorEventRepository.countBySessionIdAndEventType(sessionId, "LOOKING_DOWN"));

        List<ProctorEvent> allEvents = proctorEventRepository.findBySessionId(sessionId);
        summary.put("totalWarnings", allEvents.size());

        return summary;
    }

    private String getDefaultDescription(String eventType) {
        return switch (eventType) {
            case "TAB_SWITCH" -> "User switched to another tab or window";
            case "NO_FACE" -> "No face detected in camera view";
            case "MULTIPLE_FACES" -> "Multiple faces detected in camera view";
            case "LOOKING_AWAY" -> "User is looking away from screen";
            case "LOOKING_DOWN" -> "User is looking down frequently";
            case "BACKGROUND_NOISE" -> "Background noise or multiple voices detected";
            default -> "Suspicious activity detected";
        };
    }

    private String getWarningMessage(String eventType) {
        return switch (eventType) {
            case "TAB_SWITCH" -> "⚠️ WARNING: Tab switching detected! Stay on the interview page.";
            case "NO_FACE" -> "⚠️ WARNING: Face not detected! Please face the camera.";
            case "MULTIPLE_FACES" -> "🚨 ALERT: Multiple faces detected! Only you should be visible.";
            case "LOOKING_AWAY" -> "⚠️ WARNING: Please look at the screen.";
            case "LOOKING_DOWN" -> "⚠️ NOTICE: Frequent looking down detected.";
            case "BACKGROUND_NOISE" -> "⚠️ NOTICE: Background noise detected. Move to a quieter place.";
            default -> "⚠️ Suspicious activity detected.";
        };
    }
}

File: backend/src/services/proctoringService.js
/**
 * proctoringService.js — Proctoring event logging and integrity scoring.
 * Direct port of Java ProctoringService.
 */

const ProctorEvent = require('../models/ProctorEvent');
const InterviewSession = require('../models/InterviewSession');

// Deduction values for each event type (same as Java)
const DEDUCTIONS = {
  TAB_SWITCH: 10,
  NO_FACE: 5,
  MULTIPLE_FACES: 15,
  LOOKING_AWAY: 3,
  LOOKING_DOWN: 2,
  BACKGROUND_NOISE: 2
};

/**
 * Log a proctoring event and update integrity score
 */
async function logEvent(data) {
  // Save the event
  const event = new ProctorEvent({
    sessionId: data.sessionId,
    eventType: data.eventType,
    severity: data.severity || 'MEDIUM',
    description: data.description || getDefaultDescription(data.eventType)
  });

  await event.save();

  // Update session integrity score
  const session = await InterviewSession.findById(data.sessionId);
  if (!session) throw new Error('Session not found');

  const deduction = DEDUCTIONS[data.eventType] || 2;
  const newScore = Math.max(0, session.integrityScore - deduction);
  session.integrityScore = newScore;
  session.warningCount = session.warningCount + 1;
  await session.save();

  // Build response
  return {
    integrityScore: newScore,
    warningCount: session.warningCount,
    warning: getWarningMessage(data.eventType),
    deduction
  };
}

/**
 * Get integrity score for a session
 */
async function getIntegrityScore(sessionId) {
  const session = await InterviewSession.findById(sessionId);
  if (!session) throw new Error('Session not found');

  return {
    integrityScore: session.integrityScore,
    warningCount: session.warningCount
  };
}

/**
 * Calculate integrity score from events (used for reports)
 */
function calculateIntegrity(tabSwitches, noFace, multipleFaces, lookingAway) {
  const deduction = (tabSwitches * 10) + (noFace * 5) + (multipleFaces * 15) + (lookingAway * 3);
  return Math.max(0, 100 - deduction);
}

/**
 * Get proctoring summary for report
 */
async function getProctoringSummary(sessionId) {
  const tabSwitches = await ProctorEvent.countDocuments({ sessionId, eventType: 'TAB_SWITCH' });
  const noFaceCount = await ProctorEvent.countDocuments({ sessionId, eventType: 'NO_FACE' });
  const multipleFaces = await ProctorEvent.countDocuments({ sessionId, eventType: 'MULTIPLE_FACES' });
  const lookingAway = await ProctorEvent.countDocuments({ sessionId, eventType: 'LOOKING_AWAY' });
  const lookingDown = await ProctorEvent.countDocuments({ sessionId, eventType: 'LOOKING_DOWN' });

  const allEvents = await ProctorEvent.find({ sessionId });
  const totalWarnings = allEvents.length;

  return {
    tabSwitches,
    noFaceCount,
    multipleFaces,
    lookingAway,
    lookingDown,
    totalWarnings
  };
}

function getDefaultDescription(eventType) {
  const descriptions = {
    TAB_SWITCH: 'User switched to another tab or window',
    NO_FACE: 'No face detected in camera view',
    MULTIPLE_FACES: 'Multiple faces detected in camera view',
    LOOKING_AWAY: 'User is looking away from screen',
    LOOKING_DOWN: 'User is looking down frequently',
    BACKGROUND_NOISE: 'Background noise or multiple voices detected'
  };
  return descriptions[eventType] || 'Suspicious activity detected';
}

function getWarningMessage(eventType) {
  const warnings = {
    TAB_SWITCH: '⚠️ WARNING: Tab switching detected! Stay on the interview page.',
    NO_FACE: '⚠️ WARNING: Face not detected! Please face the camera.',
    MULTIPLE_FACES: '🚨 ALERT: Multiple faces detected! Only you should be visible.',
    LOOKING_AWAY: '⚠️ WARNING: Please look at the screen.',
    LOOKING_DOWN: '⚠️ NOTICE: Frequent looking down detected.',
    BACKGROUND_NOISE: '⚠️ NOTICE: Background noise detected. Move to a quieter place.'
  };
  return warnings[eventType] || '⚠️ Suspicious activity detected.';
}

module.exports = { logEvent, getIntegrityScore, calculateIntegrity, getProctoringSummary };
